pipelines/SARS-CoV-2-pipelines/run_python_scripts/rules/count_observed_counts.py

```diff
@@ -0,0 +1,126 @@
+import argparse
+import sys
+import os
+
+import pysam
+
+from helpers import load_fasta, apply_to_cigartuples, create_barcodes_dict, dump_dict_to_file, load_dict
+
+letters = "ACGT"
+l2n = {letter: num for num, letter in enumerate(letters)}
+
+def parse_args(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument("reference", type=str)
+    parser.add_argument("-o", "--output", type=str, required=True)
+    parser.add_argument("--alignment_low_cutoff", type=int, default=50)
+    parser.add_argument("--working_path", type=str, required=True)
+    parser.add_argument("--fastq_path", type=str, required=True)
+    parser.add_argument("--rampart_csv_path", type=str, required=True)
+    return parser.parse_args(argv)
+
+def create_bam(reference,fname,fastq_path,working_path):
+    print("minimap2 -t 2 -x map-ont -a "+reference+" "+fastq_path+"/"+fname+".fastq | samtools view -S -b -o - | samtools sort - -o "+working_path+"/"+fname+".bam && samtools index "+working_path+"/"+fname+".bam")
+    res=os.system("minimap2 -t 2 -x map-ont -a "+reference+" "+fastq_path+"/"+fname+".fastq | samtools view -S -b -o - | samtools sort - -o "+working_path+"/"+fname+".bam && samtools index "+working_path+"/"+fname+".bam" )

```

Comment:
os.system is dangerous as it invokes shell + this does not handle escaping arguments well. prefer `subprocess.run(['binary', 'list', 'of', 'arguments'])

Reply:
hmm, something like this?
```
p1 = subprocess.Popen(["minimap2", "-t 2", "-x", "map-ont", "-a", reference, fastq_path+"/"+fname+".fastq" ], stdout=subprocess.PIPE)
p2 = subprocess.Popen(["samtools", "view", "-S", "-b", "-o", "-" ], stdin=p1.stdout, stdout=subprocess.PIPE)
subprocess.run(["samtools", "sort", "-", "-o "+working_path+"/"+fname+".bam" ], stdin=p2.stdout)
subprocess.run(["samtools", "index", working_path+"/"+fname+".bam"])
```
I am not sure how to implement it correctly using subprocess
(I've tried this, but it did not create the .bam file, but i got no error)